### ISSUE TITLE
Linter should suggest having a space on either side of any binary operators, and before any unary operators

### ISSUE THREAD

**Issue by Hixie** [dart-lang/linter#249]
This code shows why you might care:
```dart
void main() {
  print(5 ~/ 10); // prints the result of integer divining five by ten
  print(5 /~ 10); // prints the result of dividing five by the bitwise-complement of ten
  int x = 10;
  x += 1; // adds one to x
  x =+ 1; // sets x to one
}
```

With the suggested lint, the following code would be ok:
```dart
void main() {
  print(5 ~/ 10);
  print(5 / ~10);
  int x = 10;
  x += 1;
  x = +1;
}
```

**Comment by zoechi** [comment]
Shouldn't this be handled by dartfmt?

**Comment by pq** [comment]
> Shouldn't this be handled by dartfmt?

In general, yes.  However, some folks don't use it.  (Specifically, [flutter](https://flutter.io/style-guide/#formatting).)

**Comment by pq** [comment]
FWIW this bit 
```dart
x =+ 1; // sets x to one
```
produces a warning (_"expected an identifier"_).  Maybe it didn't before?

@Hixie : is the intent here to make it more obvious when someone intends a binary op but is getting a unary one?  Typing, for example, `5 /~ 10` when they meant `5 ~/ 10`?

**Comment by bwilkerson** [comment]
The code `x =+ 1;` produces an error because `+` is not a valid prefix operator.

**Comment by Hixie** [comment]
> is the intent here to make it more obvious when someone intends a binary op but is getting a unary one? Typing, for example, `5 /~ 10` when they meant `5 ~/ 10`?

That was in fact the exact case that led to my filing this bug.

Regarding `=+`, change all my examples from using `+` to using `-`.

**Comment by pq** [comment]
**Update**: the lint as described in 1fdc9f9 caused enough confusion (namely around the overlapping roles of the formatter and analyzer) that we've decided to back it out pending more consideration.  Along those lines, @bwilkerson had some interesting thoughts around focusing on the particular mistakes (e.g., `~/` vs. `/~`) rather than more generally linting something ensured by the formatter.

_Sidenote: this further emphasizes the value in finding a happy place where the formatter plays nice with (in this case Flutter) idiomatic Dart.  Luckily, folks are working on just that._ 👍

**Comment by Hixie** [comment]
If we had a formatter that didn't mess up expressions like:

```dart
  if (offset.x == velocity.x * xFactor + x0 &&
      offset.y == velocity.y           + xy)
    return true;
```

...then sign me up. Until then, a linter can solve the problem described in the OP in a way a formatter can't.

**Comment by zoechi** [comment]
@Hixie I guess in this case I'm in favor of linter rules as well :D
